ports-of-call/portable_errors.hpp

```diff
@@ -0,0 +1,209 @@
+// so make sure you allocate enough memory.
+#define PORTABLE_ERROR_MESSAGE(message, output)                                \
+  PortsOfCall::ErrorChecking::error_msg(message, __FILE__, __LINE__, output);
+
+// Aborts the program with an error message with file and line number.
+// if PORTABILITY_STRATEGY_NONE is enabled, then this throws a C++
+// exception. Otherwise exceptions are not used.
+#ifdef PORTABILITY_STRATEGY_NONE
+#define PORTABLE_ALWAYS_THROW_OR_ABORT(message)                                \
+  PortsOfCall::ErrorChecking::abort_throws(message, __FILE__, __LINE__);
+#else
+#define PORTABLE_ALWAYS_THROW_OR_ABORT(message)                                \
+  PortsOfCall::ErrorChecking::abort(message, __FILE__, __LINE__);
+#endif // PORTABILITY_STRATEGY_NONE
+
+// Same as the above, but disabled when compiled in release mode.
+#ifdef NDEBUG
+#define PORTABLE_REQUIRE(condition, message) ((void)0)
+#define PORTABLE_ABORT(message) ((void)0)
+#define PORTABLE_WARN(message) ((void)0)
+#define PORTABLE_THROW_OR_ABORT(message) ((void)0)
+#else
+#define PORTABLE_REQUIRE(condition, message)                                   \
+  PORTABLE_ALWAYS_REQUIRE(condition, message)
+#define PORTABLE_ABORT(message) PORTABLE_ALWAYS_ABORT(message)
+#define PORTABLE_WARN(message) PORTABLE_ALWAYS_WARN(message)
+#define PORTABLE_THROW_OR_ABORT(message) PORTABLE_ALWAYS_THROW_OR_ABORT(message)
+#endif // NDEBUG
+
+// Below are the underlying type-safe functions called by the
+// macros. The functions are not to be called by downstream users, as
+// they don't know file and line numbers. They simply enable type
+// checking for the above macros,
+namespace PortsOfCall {
+namespace ErrorChecking {
+namespace impl {
+// Abort with an error message. The abort function depends on
+// portability backend.
+[[noreturn]] PORTABLE_INLINE_FUNCTION void abort(const char *const message) {
+#ifdef PORTABILITY_STRATEGY_KOKKOS
+  Kokkos::abort(message);
+  // For some versions of Kokkos, Kokkos::abort ends control flow, but
+  // is not marked as `[[noreturn]]`, so we need this loop to supress
+  // a warning that the function does not return.
+  while (true) {
+  }
+#else
+  std::abort();
+#endif // PORTABILITY_STRATEGY_KOKKOS
+}
+} // namespace impl
+
+// Prints an error message describing the failed condition in an
+// assert (the assertion is applied above in the macro) with file name
+// and line number. Then aborts.
+[[noreturn]] PORTABLE_INLINE_FUNCTION void require(const char *const condition,
+                                                   const char *const message,
+                                                   const char *const filename,
+                                                   int const linenumber) {
+  std::printf("### ERROR\n  Condition:   %s\n  Message:     %s\n  File:        "
+              "%s\n  Line number: %i\n",
+              condition, message, filename, linenumber);
+  impl::abort(message);
```

Comment:
won't this double print the message? same for below

Reply:
yes it does---personally I think this is fine, but if it bothers you I can tweak to avoid that

Reply:
If this happens to get used at a low level, it could get very annoying (ie **twice** per material, per iteration (if pte solve), per cell, per rank, per time step). Could you just pass in an empty string?

Reply:
Yeah I'll do that.